### PR TITLE
Add internal snapshot_and json pseudo protocol cases

### PR DESCRIPTION
**Pull request by chunfuwen**
Ceph support internal snapshot,and attaching front end disk created
by json pseudo protocol

Signed-off-by: chunfuwen <chwen@redhat.com>

### CODE REVIEW

**Comment by chunfuwen** [comment]
@will-Do 
@balamuruhans 

Please help take a look

**Comment by chunfuwen** [comment]
@balamuruhans 

Could you please take a look?

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 1]
```diff
@@ -143,7 +144,7 @@ def check_vol(vol_params):
         libvirt.check_exit_status(ret)
         # vol-info
         if not pv.volume_info(vol_name):
-            raise test.fail("Can't see volmue info")
+            raise test.fail("Can't see volume info")
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: `raise` is not required

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 2]
```diff
@@ -370,6 +374,40 @@ def check_in_vm(vm_obj, target, old_parts, read_only=False):
...
+        Get all snapshots for rbd_vol.img volume,unprotect and then clean up them.
+        """
+        cmd = ("rbd -m {0} {1} info {2}"
+               "".format(mon_host, key_opt, "%s/%s" % (disk_src_pool, vol_name)))
+        if process.run(cmd, ignore_status=True, shell=True).exit_status:
+            return
+        # Get snapshot list.
+        cmd = ("rbd -m {0} {1} snap"
+               " list {2}"
+               "".format(mon_host, key_opt, "%s/%s" % (disk_src_pool, vol_name)))
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: `os.path.join(disk_src_pool, vol_name)`

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 3]
```diff
@@ -370,6 +374,40 @@ def check_in_vm(vm_obj, target, old_parts, read_only=False):
...
+        snaps_out = process.run(cmd, ignore_status=True, shell=True).stdout
+        snap_names = []
+        if snaps_out:
+            for line in snaps_out.rsplit("\n"):
+                if line.startswith("SNAPID") or line == "":
+                    continue
+                snap_line = line.rsplit()
+                if len(snap_line) == 4:
+                    snap_names.append(snap_line[1])
+            logging.debug("Find snapshots: %s" % snap_names)
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: `logging.debug("Find snapshots: %s", snap_names)`

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 4]
```diff
@@ -370,6 +374,40 @@ def check_in_vm(vm_obj, target, old_parts, read_only=False):
...
+            # Unprotect snapshot first,otherwise it will fail to purge volume
+            for snap_name in snap_names:
+                cmd = ("rbd -m {0} {1} snap"
+                       " unprotect {2}@{3}"
+                       "".format(mon_host, key_opt, "%s/%s" % (disk_src_pool, vol_name), snap_name))
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: os.path.join can be used

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 5]
```diff
@@ -370,6 +374,40 @@ def check_in_vm(vm_obj, target, old_parts, read_only=False):
...
+                process.run(cmd, ignore_status=True, shell=True)
+        # Purge volume,and then delete volume.
+        cmd = ("rbd -m {0} {1} info {2} && rbd -m {0} {1} snap"
+               " purge {2} && rbd -m {0} {1} rm {2}"
+               "".format(mon_host, key_opt, "%s/%s" % (disk_src_pool, vol_name)))
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: same here

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 6]
```diff
@@ -667,6 +732,14 @@ def check_in_vm(vm_obj, target, old_parts, read_only=False):
                    " purge {2} && rbd -m {0} {1} rm {2}"
                    "".format(mon_host, key_opt, disk_src_name))
             process.run(cmd, ignore_status=True, shell=True)
+        elif create_volume:
+            cmd = ("rbd -m {0} {1} info {2} && rbd -m {0} {1} rm {2}"
+                   "".format(mon_host, key_opt, "%s/%s" % (disk_src_pool, cloned_vol_name)))
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: same here

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 7]
```diff
@@ -667,6 +732,14 @@ def check_in_vm(vm_obj, target, old_parts, read_only=False):
...
+            process.run(cmd, ignore_status=True, shell=True)
+            cmd = ("rbd -m {0} {1} info {2} && rbd -m {0} {1} rm {2}"
+                   "".format(mon_host, key_opt, "%s/%s" % (disk_src_pool, create_from_cloned_volume)))
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: here too

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 1 above)
Comment (chunfuwen) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: done

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 2 above)
Comment (chunfuwen) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: done

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 3 above)
Comment (chunfuwen) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: done

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 4 above)
Comment (chunfuwen) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: done

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 5 above)
Comment (chunfuwen) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: done

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 6 above)
Comment (chunfuwen) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: done

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 7 above)
Comment (chunfuwen) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: done

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 8]
```diff
@@ -635,16 +700,16 @@ def check_in_vm(vm_obj, target, old_parts, read_only=False):
             session = vm.wait_for_login()
             new_parts = libvirt.get_parts_list(session)
             if len(new_parts) != len(old_parts):
-                raise test.fail("Disk still exists in vm"
+                test.fail("Disk still exists in vm"
                                 " after detachment")
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: please fix pep8 indentation for this line and make travis to pass.

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 9]
```diff
@@ -635,16 +700,16 @@ def check_in_vm(vm_obj, target, old_parts, read_only=False):
...
 
     except virt_vm.VMStartError, details:
         for msg in unsupported_err:
             if msg in str(details):
-                raise test.skip(details)
+                test.skip(details)
         else:
-            raise test.fail("VM failed to start."
+            test.fail("VM failed to start."
                             "Error: %s" % str(details))
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: same here for travis to pass

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 8 above)
Comment (chunfuwen) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: done

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 9 above)
Comment (chunfuwen) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: done

**Comment by chunfuwen** [comment]
@balamuruhans 
pep8 issue had been fixed, thanks

**Comment by balamuruhans** [comment]
Thanks! @will-Do @chunfuwen

**Pull request merged**